src/System.IO/src/System/IO/Stream.cs

```diff
@@ -138,11 +139,18 @@ public virtual Task CopyToAsync(Stream destination, int bufferSize, Cancellation
             Debug.Assert(CanRead);
             Debug.Assert(destination.CanWrite);
 
-            byte[] buffer = new byte[bufferSize];
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(bufferSize);
             int bytesRead;
-            while ((bytesRead = await ReadAsync(buffer, 0, buffer.Length, cancellationToken).ConfigureAwait(false)) != 0)
+            try
             {
-                await destination.WriteAsync(buffer, 0, bytesRead, cancellationToken).ConfigureAwait(false);
+                while ((bytesRead = await ReadAsync(buffer, 0, buffer.Length, cancellationToken).ConfigureAwait(false)) != 0)
+                {
+                    await destination.WriteAsync(buffer, 0, bytesRead, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
```

Comment:
Note that many of the files being edited in this change are used for CoreRT but not for CoreCLR, e.g. when compiled for coreclr, System.IO.dll type forwards Stream to the one in mscorlib.  That means that these optimizations will accrue to CoreRT but not to CoreCLR.